src/tools/pkcs15-init.c

```diff
@@ -1340,7 +1344,10 @@ do_read_check_certificate(sc_pkcs15_cert_t *sc_oldcert,
 	r = SC_ERROR_INVALID_ARGUMENTS;
 	if (oldpk_type == newpk_type)
 	{
-#if  OPENSSL_VERSION_NUMBER >= 0x10002000L
+#if OPENSSL_VERSION_NUMBER >= 0x30000000L
+		if (EVP_PKEY_eq(oldpk, newpk))
+			r = 0;
+#elif  OPENSSL_VERSION_NUMBER >= 0x10002000L
```

Comment:
I am wondering if this would be a good candidate for a compat function as the functions look quite similar.